### DISCUSSION TITLE
Rendering Angular component inside of the jointjs element (e.g. rectangle)

### DISCUSSION THREAD

**Discussion by silviodjuric** [clientIO/joint#2253] [Q&A]
### Introduction

So I have created my graph and I alreday have elements on it with node connections and other logic. Currently I'm working on rendering Angular component (search with autocomplete) that is already implemented in my JointJs graph.

I actually want to render it in a rectangle element and then have it placed on the paper and use it with my other nodes.

Does anyone have any idea how can this be done? Prefferable way would be if I could do it through .ts file rather than doing something like rendering foreignObject in html template.

I have tried something like (code in the comments):

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS, JointJS+

**Comment by silviodjuric** [comment]
```
 @ViewChild('searchTemplate') searchTemplate: TemplateRef<any>;
  @ViewChild('paper') paperElement: ElementRef;

  ngAfterViewInit() {
    const paperContainer = this.paperElement.nativeElement;

    // Create and append the rectangle element
    const rectangleElement = this.renderer.createElement('rect');
    this.renderer.setAttribute(rectangleElement, 'x', 'x-coordinate');
    this.renderer.setAttribute(rectangleElement, 'y', 'y-coordinate');
    this.renderer.setAttribute(rectangleElement, 'width', 'width');
    this.renderer.setAttribute(rectangleElement, 'height', 'height');
    this.renderer.setStyle(rectangleElement, 'fill', 'color');
    this.renderer.setAttribute(rectangleElement, 'draggable', 'true');
    this.renderer.appendChild(paperContainer, rectangleElement);

    // Create and append the foreignObject with the search component
    const foreignObjectElement = this.renderer.createElement('foreignObject');
    this.renderer.setAttribute(foreignObjectElement, 'x', 'x-coordinate');
    this.renderer.setAttribute(foreignObjectElement, 'y', 'y-coordinate');
    this.renderer.setAttribute(foreignObjectElement, 'width', 'width');
    this.renderer.setAttribute(foreignObjectElement, 'height', 'height');
    const searchContainer = this.renderer.createElement('div');
    this.renderer.setAttribute(searchContainer, 'class', 'search-container');
    const searchTemplate = this.searchTemplate.createEmbeddedView(null);
    this.renderer.appendChild(searchContainer, searchTemplate.rootNodes[0]);
    this.renderer.appendChild(foreignObjectElement, searchContainer);
    this.renderer.appendChild(paperContainer, foreignObjectElement);

    // Add dragstart event listener to the rectangle for dragging
    rectangleElement.addEventListener('dragstart', (event) => {
      event.dataTransfer.setData('text', ''); // Enable dragging in some browsers
    });
  }
```

```
<div #toolbar></div>
<div class="main-container">
  <div class="paper-container" #paper>
    <svg class="flowchart-svg">
      <!-- Other flowchart elements -->
    </svg>
  </div>
</div>

<ng-template #searchTemplate>
  <div class="search-container">
    <tt-component-search></tt-component-search>
  </div>
</ng-template>
```

**Comment by kumilingus** [comment]
If you want to define the element structure by using the _DOM API_ (not declaratively) you can check this [tutorial](https://resources.jointjs.com/tutorial/content-driven-element) or these demos:
- https://www.jointjs.com/demos/microsoft-adaptive-cards
- https://www.jointjs.com/demos/counters

If you're looking for tighter integration with Angular, you might want to look at a third-party project like this one [one](https://github.com/dgwnu/ng-joint/blob/master/README.md).

**Comment by coyoteecd** [comment]
@silviodjuric, you do this the other way around. Namely, your goal is to render a paper onto your own Angular component, that potentially has other child components as well (such as a search component). Your paper container component would look like this:

```
<div id="paperScroller">
  <my-search-component />
  <div id="paper"></div>
</div>
```
and in the ts file you create the paper passing the *existing* DOM element selector (so that JointJS renders itself onto the existing div elements):
```ts
    this.paper = new dia.Paper({
      el: '#paper',
      model: this.graph,
    });
    this.paperScroller = new ui.PaperScroller({
      el: '#paperScroller',
      paper: this.paper,
    });
```

@kumilingus I think this approach, or something similar, should be mentioned in the [Angular tutorial](https://resources.jointjs.com/tutorial/angular), it took me a while to figure out the library has no problem using existing elements. The info exists in the docs (the example in Paper docs shows it), but what happens in practice is that you'll follow the tutorial to get going.

**Reply by kumilingus** [reply]
I am not sure I understand. What does the `<my-search-component>` do? And why is it child of the paper scroller?

I thought that @silviodjuric wanted to use an existing Angular component inside of JointJS shape.

**Reply by coyoteecd** [reply]
Nevermind, I might have misunderstood what he's asking (thought he has trouble showing angular components as children of the paper, not of paper elements)
